### PR TITLE
CAN driver

### PR DESCRIPTION
**Pull request by haitomatic**
## Summary
- CAN driver for interfacing with FPGA CAN IP Block supporting both classical CAN and CAN-FD
## Impact
- Networking / SocketCAN
## Testing
- Tested with PX4 and Saluki v1,v2
- ESC control flight test result: https://ssrc.atlassian.net/wiki/spaces/~626f763fa1de4100684ec2ec/pages/559939795/Nuttx+CAN+FD+driver#CAN-DRONE-TEST-w-UAVCAN-v0

### CODE REVIEW

File: sched/semaphore/sem_trywait.c [hunk 1]
```diff
@@ -72,7 +72,7 @@ int nxsem_trywait(FAR sem_t *sem)
 
   /* This API should not be called from interrupt handlers & idleloop */
 
-  DEBUGASSERT(sem != NULL && up_interrupt_context() == false);
```
Comment (haitomatic) [inline sched/semaphore/sem_trywait.c:75]: Relaxing overly strict assertion. trywait does not block so it's safe to be called from interrupt handlers. https://www.mail-archive.com/commits@nuttx.apache.org/msg44549.html

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 2]
```diff
@@ -0,0 +1,3181 @@
...
+
+      /* Clear Data Overrun interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_DOI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Clear and reset all interrupt. */
+
+  caninfo("Reset all interrupts...\n");
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:1646]: I don't quite understand why the interrupt routine has to manually disable the interrupt source, will this not result in missed interrupts ? Where does the source get re-enabled ?

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 3]
```diff
@@ -0,0 +1,3181 @@
...
+ *  None
+ *
+ ****************************************************************************/
+
+static bool mpfs_can_is_txt_buf_writable(FAR struct mpfs_driver_s *priv,
+                                         uint8_t buf)
+{
+  enum mpfs_can_txtb_status buf_status;
+
+  buf_status = mpfs_can_get_tx_status(priv, buf);
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:1705]: Very difficult to understand what happens here, so do you have like n-buffers for TX, and you need to see per index if a buffer has been handled or no ? 

Why not just use a freelist for free buffers and a sendlist for buffers that need sending. Another option is a FIFO. Maybe this is just some detail with the CAN IP I don't understand, but like I said, this is not immediately obvious why the per-buffer states are needed ?

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 4]
```diff
@@ -0,0 +1,3181 @@
...
+            break;
+          }
+        priv->can.bittiming.sample_point =
+          req->arbi_samplep * 10; /* In one-tenth of a percent */
+
+        if (req->data_bitrate > 4000 ||
+            req->data_bitrate < req->arbi_bitrate)
+          {
+            canerr("Data bitrate higher than 4Mbps is yet to be supported. "
+                   "Data br cannot be smaller than arbitration br.");
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:2864]: Why not? (Does not make sense as SNR for arbitration will be better, but why is this not possible)?

File: drivers/can/can.c [hunk 5]
```diff
@@ -556,7 +556,7 @@ static ssize_t can_read(FAR struct file *filep, FAR char *buffer,
   FAR struct can_dev_s     *dev = inode->i_private;
 #endif
 
-  caninfo("buflen: %d\n", buflen);
+  caninfo("buflen: %d\n", (int)buflen);
```
Comment (pussuw) [inline drivers/can/can.c:559]: Put these changes to another PR and upstream it.

Also, size_t is unsigned int, so the correct way to do this would be

caninfo("buflen: %lu\n", (unsigned long)buflen);

Unsigned long guarantees that whatever sized size_t is, the print works.

File: sched/semaphore/sem_trywait.c (hunk 1 above)
Comment (pussuw) [reply sched/semaphore/sem_trywait.c:75]: Why would interrupts need semaphore locking ? Semaphores are a user space API.

More specifically, what should the ISR procedure that needs a semaphore lock do, in case the semaphore is locked ? I assume it was going to do something but was prevented from doing it ?

**Comment by haitomatic** [comment]
I will close this PR for now as there needs to be some modification to remove the GPL v2.0 codes which is not compatible with APACHE 2.0 Nuttx license.  Thanks @pussuw , I will take into account your suggestions in the next PR.

File: drivers/can/can.c (hunk 5 above)
Comment (jlaitine) [reply drivers/can/can.c:559]: C99 is commonly used nowdays and inttypes.h. So you can just change "%d" into "%zu" IMHO

File: sched/semaphore/sem_trywait.c (hunk 1 above)
Comment (jlaitine) [reply sched/semaphore/sem_trywait.c:75]: same same. I think this debugassert is valid, it is a bug to call this from isr. Where does this bite?

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c [hunk 6]
```diff
@@ -0,0 +1,3181 @@
...
+# define do_div(n, base) ({       \
+  uint32_t __base = (base);       \
+  uint32_t __rem;         \
+  __rem = ((uint64_t)(n)) % __base;   \
+  (n) = ((uint64_t)(n)) / __base;     \
+  __rem;            \
+ })
+
+#define MPFS_CAN_FD_TXTNF(priv) \
+  (!!(getreg32(priv->base + MPFS_CANFD_STATUS_OFFSET) & MPFS_CANFD_STATUS_TXNF))
```
[the remark this answers is not in the document]
Comment (jlaitine) [reply arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:193]: This is an old hacker's way to say "!=0", i.e. change to 0 to boolean false, others to boolean true. I'd prefer just writing "!=0", this is more common nowdays when trickery is no longer appreciated :)

File: arch/risc-v/src/mpfs/mpfs_fpga_canfd.c (hunk 3 above)
Comment (jlaitine) [reply arch/risc-v/src/mpfs/mpfs_fpga_canfd.c:1705]: Is this similar to ethernet mac dma; i.e. the HW scans for the buffers and sends them out, after which it writes the buffer status. And SW just checks if the buffer has already been sent out (i.e. it is free to write new data into)?

File: drivers/can/can.c (hunk 5 above)
Comment (pussuw) [reply drivers/can/can.c:559]: Won't work with NuttX base code, which is ANSI C

File: sched/semaphore/sem_trywait.c (hunk 1 above)
Comment (haitomatic) [reply sched/semaphore/sem_trywait.c:75]: Hi, it's needed in an ISR where a frame is needed to be sent to the socketCAN interface so that higher level user space application can receive the RX packet (via callback). In this `can_input` func, it needs to call `can_callback` to notify the user application listener about the RX packet. There it needs to lock, send event and unlock the network.
https://github.com/tiiuae/nuttx/blob/c64610f711eb7881808f7d71ea5a1f64d5807137/net/can/can_input.c#L176
```
      /* Send to socket interface */

      NETDEV_RXPACKETS(&priv->dev);
      can_input(&priv->dev);
```

Comment (haitomatic) [reply sched/semaphore/sem_trywait.c:75]: ofc, this `can_input` for sending can frame to user application can be delegated to a seperate work queue but for high performance, ideally the RX frame needs to be forwarded upstream to user application asap there is ISR triggered for RX.
And it's programmed that way in all other SocketCAN compatible CAN driver in Nuttx
For ex:
https://github.com/tiiuae/nuttx/blob/c64610f711eb7881808f7d71ea5a1f64d5807137/arch/arm/src/stm32h7/stm32_fdcan_sock.c#L1197

Comment (jlaitine) [reply sched/semaphore/sem_trywait.c:75]: Hi, this looks like awful lots of code to be called from an ISR anyway. Maybe it would be better to schedule a work queue item from the ISR, from which you can then call the can_input ?

Comment (jlaitine) [reply sched/semaphore/sem_trywait.c:75]: please try to schedule a work queue item into high priority wq. I don't think it can even work properly when called directly from the ISR, it tries to allocate IOBs in some branches of the code later one.

Comment (haitomatic) [reply sched/semaphore/sem_trywait.c:75]: anyway, I can try putting the receiving functionality to a separate work queue function and test if it would compromise the performance
